File: .NET/Demo_Common/Entity/ApHeartbeat.cs
using Demo_Common.Enum;
using MessagePack;

namespace Demo_Common.Entity
{
    /// <summary>
    /// AP heartbeat data
    /// </summary>
    [MessagePackObject]
    public class ApHeartbeat
    {
        /// <summary>
        /// AP ID
        /// </summary>
        [Key(0)]
        public string Id { get; set; } = string.Empty;
        /// <summary>
        /// Config version
        /// </summary>
        [Key(1)]
        public int ConfigVersion { get; set; }
        /// <summary>
        /// Ap version
        /// </summary>
        [Key(2)]
        public string ApVersion { get; set; } = "0.0.0";
        /// <summary>
        /// Mod version
        /// </summary>
        [Key(3)]
        public string ModVersion { get; set; } = string.Empty;
        /// <summary>
        /// Message
        /// </summary>
        [Key(4)]
        public MessageCode Message { get; set; } = MessageCode.OK;
        /// <summary>
        /// External message
        /// </summary>
        [Key(5)]
        public string MessageEx { get; set; } = string.Empty;
        /// <summary>
        /// Total count in cache
        /// </summary>
        [Key(6)]
        public int WaitCount { get; set; }
        /// <summary>
        /// Current sending count
        /// </summary>
        [Key(7)]
        public int SendCount { get; set; }
        /// <summary>
        /// Tag results list
        /// </summary>
        [Key(8)]
        public List<TagHeartbeat> Tags { get; set; } = new List<TagHeartbeat>();
    }
}

File: Python/Entities/apHeartbeat.py
import msgpack
from dataclasses import dataclass, field
from typing import List
from Enums.messageCode import MessageCodes
from .tagHeartbeat import TagHeartbeat

@dataclass
class ApHeartbeat:
    """
    AP heartbeat data
    """
    Id: str = ""
    ConfigVersion: int = 0
    ApVersion: str = "0.0.0"
    ModVersion: str = ""
    Message: MessageCodes = MessageCodes.Ok
    MessageEx: str = ""
    WaitCount: int = -1
    SendCount: int = 0
    Tags: List[TagHeartbeat] = field(default_factory=list)

    def to_msgpack(self) -> bytes:
        data = [
            self.Id,
            self.ConfigVersion,
            self.ApVersion,
            self.ModVersion,
            self.Message.value,
            self.MessageEx,
            self.WaitCount,
            self.SendCount,
            [tag.to_msgpack() for tag in self.Tags]
        ]
        return msgpack.packb(data)

    @staticmethod
    def from_msgpack(packed: bytes):
        data = msgpack.unpackb(packed)
        tags = [TagHeartbeat.from_msgpack(tag_bytes) for tag_bytes in data[8]]
        return ApHeartbeat(
            Id=data[0],
            ConfigVersion=data[1],
            ApVersion=data[2],
            ModVersion=data[3],
            Message=MessageCodes(data[4]),
            MessageEx=data[5],
            WaitCount=data[6],
            SendCount=data[7],
            Tags=tags
        )